File: csharp/Betfair.ESAClient/Betfair.ESAClient/Protocol/RequestResponseProcessor.cs
using Betfair.ESAClient.Cache;
using Betfair.ESASwagger.Model;
using Newtonsoft.Json;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Betfair.ESAClient.Auth;

namespace Betfair.ESAClient.Protocol
{
    /// <summary>
    /// This class implements the basic request / response handling of the protocol
    /// </summary>
    public class RequestResponseProcessor
    {
        public const string OPERATION = "op";

        public const string REQUEST_AUTHENTICATION = "authentication";
        public const string REQUEST_MARKET_SUBSCRIPTION = "marketSubscription";
        public const string REQUEST_ORDER_SUBSCRIPTION = "orderSubscription";
        public const string REQUEST_HEARTBEAT = "heartbeat";

        public const string RESPONSE_CONNECTION = "connection";
        public const string RESPONSE_STATUS = "status";
        public const string RESPONSE_MARKET_CHANGE_MESSAGE = "mcm";
        public const string RESPONSE_ORDER_CHANGE_MESSAGE = "ocm";

        private int _nextId;
        private TaskCompletionSource<ConnectionMessage> _connectionMessage = new TaskCompletionSource<ConnectionMessage>();
        private ConcurrentDictionary<int, RequestResponse> _tasks = new ConcurrentDictionary<int, RequestResponse>();

        //subscription handlers
        private SubscriptionHandler<MarketSubscriptionMessage, ChangeMessage<MarketChange>, MarketChange> _marketSubscriptionHandler;
        private SubscriptionHandler<OrderSubscriptionMessage, ChangeMessage<OrderMarketChange>, OrderMarketChange> _orderSubscriptionHandler;

        private IChangeMessageHandler _changeHandler;
        private Action<string> _sendLine;
        private ConnectionStatus _status = ConnectionStatus.STOPPED;

        public DateTime LastRequestTime { get; private set; }
        public DateTime LastResponseTime { get; private set; }

        /// <summary>
        /// Lock used to syncronize the send process.
        /// </summary>
        private object _sendLock = new object();

        public RequestResponseProcessor(Action<string> sendLine)
        {
            _sendLine = sendLine;
            ChangeHandler = null;
            LastRequestTime = DateTime.MinValue;
            LastResponseTime = DateTime.MinValue;
        }

        public ConnectionStatus Status
        {
            get
            {
                return _status;
            }
            set
            {
                if (value == _status)
                {
                    //no-op
                    return;
                }
                ConnectionStatusEventArgs args = new ConnectionStatusEventArgs() { Old = _status, New = value };
                Trace.TraceInformation("ESAClient: Status changed {0} -> {1}", _status, value);
                _status = value;

                DispatchConnectionStatusChanged(args);
            }
        }

        private void DispatchConnectionStatusChanged(ConnectionStatusEventArgs args)
        {
            if(ConnectionStatusChanged != null)
            {
                ConnectionStatusChanged(this, args);
            }
        }

        /// <summary>
        /// Trace change messages but truncate to specified chars (0 is off).
        /// </summary>
        public int TraceChangeTruncation { get; set; }

        private void Reset()
        {
            _connectionMessage.TrySetCanceled();
            _connectionMessage = new TaskCompletionSource<ConnectionMessage>();
            foreach (RequestResponse task in _tasks.Values)
            {
                task.Cancelled();
            }
            _tasks = new ConcurrentDictionary<int, RequestResponse>();
        }

        public void Disconnected()
        {
            Status = ConnectionStatus.DISCONNECTED;
            Reset();
        }

        public void Stopped()
        {
            MarketSubscriptionHandler = null;
            OrderSubscriptionHandler = null;
            Status = ConnectionStatus.STOPPED;
            Reset();
        }

        public MarketSubscriptionMessage MarketResubscribeMessage
        {
            get
            {
                if(MarketSubscriptionHandler != null)
                {
                    MarketSubscriptionMessage resub = MarketSubscriptionHandler.SubscriptionMessage;
                    resub.InitialClk = MarketSubscriptionHandler.InitialClk;
                    resub.Clk = MarketSubscriptionHandler.Clk;
                    return resub;
                }
                return null;
            }
        }

        public OrderSubscriptionMessage OrderResubscribeMessage
        {
            get
            {
                if (OrderSubscriptionHandler != null)
                {
                    OrderSubscriptionMessage resub = OrderSubscriptionHandler.SubscriptionMessage;
                    resub.InitialClk = OrderSubscriptionHandler.InitialClk;
                    resub.Clk = OrderSubscriptionHandler.Clk;
                    return resub;
                }
                return null;
            }
        }


        public IChangeMessageHandler ChangeHandler
        {
            get
            {
                return _changeHandler;
            }
            set
            {
                if (_changeHandler == null)
                {
                    _changeHandler = new NullChangeHandler();
                } else
                {
                    _changeHandler = value;
                }
            }
        }

        public SubscriptionHandler<MarketSubscriptionMessage, ChangeMessage<MarketChange>, MarketChange> MarketSubscriptionHandler
        {
            get
            {
                return _marketSubscriptionHandler;
            }

            set
            {
                if (_marketSubscriptionHandler != null) _marketSubscriptionHandler.Cancel();
                _marketSubscriptionHandler = value;
                if(value != null) Status = ConnectionStatus.SUBSCRIBED;              
            }
        }

        public SubscriptionHandler<OrderSubscriptionMessage, ChangeMessage<OrderMarketChange>, OrderMarketChange> OrderSubscriptionHandler
        {
            get
            {
                return _orderSubscriptionHandler;
            }

            set
            {
                if (_orderSubscriptionHandler != null) _orderSubscriptionHandler.Cancel();
                _orderSubscriptionHandler = value;
                if (value != null) Status = ConnectionStatus.SUBSCRIBED;
            }
        }

        public Task<ConnectionMessage> ConnectionMessage()
        {
            return _connectionMessage.Task;
        }

        public Task<StatusMessage> Authenticate(AuthenticationMessage message)
        {
            message.Id = NextId();
            message.Op = REQUEST_AUTHENTICATION;
            return SendMessage(new RequestResponse((int)message.Id, message, 
                success => Status = ConnectionStatus.AUTHENTICATED));
        }

        public Task<StatusMessage> Heartbeat(HeartbeatMessage message)
        {
            message.Id = NextId();
            message.Op = REQUEST_HEARTBEAT;
            return SendMessage(new RequestResponse((int)message.Id, message, null));
        }

        public Task<StatusMessage> MarketSubscription(MarketSubscriptionMessage message)
        {
            int id = NextId();
            message.Id = id;
            message.Op = REQUEST_MARKET_SUBSCRIPTION;
            var newSub = new SubscriptionHandler<MarketSubscriptionMessage, ChangeMessage<MarketChange>, MarketChange>(id, message, false);
            return SendMessage(new RequestResponse(id, message,
                success => MarketSubscriptionHandler = newSub));
        }

        public Task<StatusMessage> OrderSubscription(OrderSubscriptionMessage message)
        {
            int id = NextId();
            message.Id = id;
            message.Op = REQUEST_ORDER_SUBSCRIPTION;
            var newSub = new SubscriptionHandler<OrderSubscriptionMessage, ChangeMessage<OrderMarketChange>, OrderMarketChange>(id, message, false);
            return SendMessage(new RequestResponse(id, message, 
                success => OrderSubscriptionHandler = newSub));
        }

        private int NextId()
        {
            int id = Interlocked.Increment(ref _nextId);
            return id;
        }


        private Task<StatusMessage> SendMessage(RequestResponse requestResponse)
        {
            lock (_sendLock)
            {
                //store a future task
                _tasks[requestResponse.Id] = requestResponse;

                //serialize message & send
                string line = JsonConvert.SerializeObject(requestResponse.Request, Formatting.None);
                Trace.TraceInformation("Client->ESA: " + line);

                //send line
                _sendLine(line);

                //time
                LastRequestTime = DateTime.UtcNow;
            }

            return requestResponse.Task;
        }

        /// <summary>
        /// Process a line of json
        /// </summary>
        /// <exception cref="JsonException">Thrown if line was invalid json</exception>
        /// <param name="line"></param>
        public ResponseMessage ReceiveLine(string line)
        {
            //clear last response
            ResponseMessage message = null;
            LastResponseTime = DateTime.UtcNow;
            var time = Stopwatch.StartNew();
            string operation = GetOperation(new JsonTextReader(new StringReader(line)));
            switch (operation)
            {
                case RESPONSE_CONNECTION:
                    Trace.TraceInformation("ESA->Client: {0}", line);
                    ConnectionMessage connectionMessage = ReadResponseMessage<ConnectionMessage>(line);
                    message = connectionMessage;
                    ProcessConnectionMessage(connectionMessage);
                    break;
                case RESPONSE_STATUS:
                    Trace.TraceInformation("ESA->Client: {0}", line);
                    StatusMessage statusMessage = ReadResponseMessage<StatusMessage>(line);
                    message = statusMessage;
                    ProcessStatusMessage(statusMessage);
                    break;
                case RESPONSE_MARKET_CHANGE_MESSAGE:
                    TraceChange(line);
                    MarketChangeMessage marketChangeMessage = ReadResponseMessage<MarketChangeMessage>(line);
                    message = marketChangeMessage;
                    ProcessMarketChangeMessage(marketChangeMessage);
                    break;
                case RESPONSE_ORDER_CHANGE_MESSAGE:
                    TraceChange(line);
                    OrderChangeMessage orderChangeMessage = ReadResponseMessage<OrderChangeMessage>(line);
                    message = orderChangeMessage;
                    ProcessOrderChangeMessage(orderChangeMessage);
                    break;
                default:
                    Trace.TraceError("ESA->Client: Unknown message type: {0}, message:{1}", operation, line);
                    break;
            }
            time.Stop();

            return message;
        }

        private void TraceChange(string line)
        {
            if (TraceChangeTruncation != 0)
            {
                Trace.TraceInformation("ESA->Client: {0}", line.Substring(0, Math.Min(TraceChangeTruncation, line.Length)));
            }              
        }

        private void ProcessOrderChangeMessage(OrderChangeMessage message)
        {
            ChangeMessage<OrderMarketChange> change = ChangeMessageFactory.ToChangeMessage(message);
            change = OrderSubscriptionHandler.ProcessChangeMessage(change);

            if (change != null) ChangeHandler.OnOrderChange(change);
        }

        private void ProcessMarketChangeMessage(MarketChangeMessage message)
        {
            ChangeMessage<MarketChange> change = ChangeMessageFactory.ToChangeMessage(message);
            change = MarketSubscriptionHandler.ProcessChangeMessage(change);

            if(change != null) ChangeHandler.OnMarketChange(change);
        }

        private void ProcessStatusMessage(StatusMessage statusMessage)
        {

            if (statusMessage.Id == null)
            {
                //async status / status for a message that couldn't be decoded
                ProcessUncorrelatedStatus(statusMessage);
            } else {
                RequestResponse task;
                _tasks.TryGetValue((int)statusMessage.Id, out task);
                if(task == null)
                {
                    //shouldn't happen
                    ProcessUncorrelatedStatus(statusMessage);
                } else
                {
                    //unwind task
                    task.ProcesStatusMessage(statusMessage);
                }
            }
        }

        public void ProcessUncorrelatedStatus(StatusMessage statusMessage)
        {
            Trace.TraceError("Error Status Notification: {0}", statusMessage);
            ChangeHandler.OnErrorStatusNotification(statusMessage);
        }

        protected void ProcessConnectionMessage(ConnectionMessage connectionMessage)
        {
            _connectionMessage.TrySetResult(connectionMessage);
            Status = ConnectionStatus.CONNECTED;
        }

        private T ReadResponseMessage<T>(string line) where T : ResponseMessage
        {
            T response = JsonConvert.DeserializeObject<T>(line);
            return response;
        }

        /// <summary>
        /// Extracts the op / operation attribute denoting the type of the message
        /// </summary>
        /// <param name="jreader">A reader containing json</param>
        /// <returns>The "op" value</returns>
        protected string GetOperation(JsonReader jreader)
        {
            string operation = null;
            if (jreader.Read())
            {
                //rip off start
                while (jreader.Read())
                {
                    if (jreader.TokenType == JsonToken.PropertyName && OPERATION.Equals(jreader.Value))
                    {
                        if (jreader.Read()) //rip out op's value
                        {
                            operation = (string)jreader.Value;
                        }
                        break;
                    }
                    else
                    {
                        jreader.Skip();
                    }
                }
            }
            return operation;
        }

        private class NullChangeHandler : IChangeMessageHandler
        {
            public void OnMarketChange(ChangeMessage<MarketChange> change)
            {
                Console.WriteLine("<null change handler>.OnMarketChange: " + change);
            }

            public void OnOrderChange(ChangeMessage<OrderMarketChange> change)
            {
                Console.WriteLine("<null change handler>.OnOrderChange: " + change);
            }

            public void OnErrorStatusNotification(StatusMessage message)
            {
                Console.WriteLine("<null change handler>.OnErrorStatusNotification: " + message);
            }
        }

        public event ConnectionStatusChangedEventHandler ConnectionStatusChanged;
    }

    public delegate void ConnectionStatusChangedEventHandler(object sender, ConnectionStatusEventArgs e);

    public class ConnectionStatusEventArgs : EventArgs
    {
        public ConnectionStatus Old { get; set; }
        public ConnectionStatus New { get; set; }
    }
}

File: java/client/src/main/java/com/betfair/esa/client/protocol/RequestResponseProcessor.java
package com.betfair.esa.client.protocol;

import com.betfair.esa.swagger.model.AuthenticationMessage;
import com.betfair.esa.swagger.model.ConnectionMessage;
import com.betfair.esa.swagger.model.HeartbeatMessage;
import com.betfair.esa.swagger.model.MarketChange;
import com.betfair.esa.swagger.model.MarketChangeMessage;
import com.betfair.esa.swagger.model.MarketSubscriptionMessage;
import com.betfair.esa.swagger.model.OrderChangeMessage;
import com.betfair.esa.swagger.model.OrderMarketChange;
import com.betfair.esa.swagger.model.OrderSubscriptionMessage;
import com.betfair.esa.swagger.model.RequestMessage;
import com.betfair.esa.swagger.model.ResponseMessage;
import com.betfair.esa.swagger.model.StatusMessage;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import java.io.IOException;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/** Created by mulveyj on 07/07/2016. */
public class RequestResponseProcessor {
    public static final String REQUEST_AUTHENTICATION = "authentication";
    public static final String REQUEST_MARKET_SUBSCRIPTION = "marketSubscription";
    public static final String REQUEST_ORDER_SUBSCRIPTION = "orderSubscription";
    public static final String REQUEST_HEARTBEAT = "heartbeat";

    public static final String RESPONSE_CONNECTION = "connection";
    public static final String RESPONSE_STATUS = "status";
    public static final String RESPONSE_MARKET_CHANGE_MESSAGE = "mcm";
    public static final String RESPONSE_ORDER_CHANGE_MESSAGE = "ocm";

    private static final Logger LOG = LoggerFactory.getLogger(RequestResponseProcessor.class);
    private final ObjectMapper objectMapper;
    private final AtomicInteger nextId = new AtomicInteger();
    private FutureResponse<ConnectionMessage> connectionMessage = new FutureResponse<>();
    private ConcurrentHashMap<Integer, RequestResponse> tasks = new ConcurrentHashMap<>();

    // subscription handlers
    private SubscriptionHandler<
                    MarketSubscriptionMessage, ChangeMessage<MarketChange>, MarketChange>
            marketSubscriptionHandler;
    private SubscriptionHandler<
                    OrderSubscriptionMessage, ChangeMessage<OrderMarketChange>, OrderMarketChange>
            orderSubscriptionHandler;

    private ChangeMessageHandler changeHandler;
    private final RequestSender sendLine;
    private ConnectionStatus status = ConnectionStatus.STOPPED;
    private final CopyOnWriteArrayList<ConnectionStatusListener> connectionStatusListeners =
            new CopyOnWriteArrayList<>();

    private long lastRequestTime = Long.MAX_VALUE;
    private long lastResponseTime = Long.MAX_VALUE;

    private int traceChangeTruncation;
    private final Object sendLock = new Object();

    public RequestResponseProcessor(RequestSender sendLine) {
        this.sendLine = sendLine;
        setChangeHandler(null);

        objectMapper = new ObjectMapper().registerModule(new JavaTimeModule());
        objectMapper.addMixIn(ResponseMessage.class, MixInResponseMessage.class);
        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    private void setStatus(ConnectionStatus value) {
        if (value == status) {
            // no-op
            return;
        }
        ConnectionStatusChangeEvent args = new ConnectionStatusChangeEvent(this, status, value);
        LOG.info("ESAClient: Status changed {} -> {}", status, value);
        status = value;

        dispatchConnectionStatusChange(args);
    }

    public ConnectionStatus getStatus() {
        return status;
    }

    private void dispatchConnectionStatusChange(ConnectionStatusChangeEvent args) {
        try {
            connectionStatusListeners.forEach(c -> c.connectionStatusChange(args));
        } catch (Exception e) {
            LOG.error("Exception during event dispatch", e);
        }
    }

    public void addConnectionStatusListener(ConnectionStatusListener listener) {
        connectionStatusListeners.add(listener);
    }

    public void removeConnectionStatusListener(ConnectionStatusListener listener) {
        connectionStatusListeners.remove(listener);
    }

    public int getTraceChangeTruncation() {
        return traceChangeTruncation;
    }

    public void setTraceChangeTruncation(int traceChangeTruncation) {
        this.traceChangeTruncation = traceChangeTruncation;
    }

    public long getLastRequestTime() {
        return lastRequestTime;
    }

    public long getLastResponseTime() {
        return lastResponseTime;
    }

    private void reset() {
        ConnectionException cancelException =
                new ConnectionException("Connection reset - task cancelled");
        connectionMessage.setException(cancelException);
        connectionMessage = new FutureResponse<>();
        for (RequestResponse task : tasks.values()) {
            task.getFuture().setException(cancelException);
        }
        tasks = new ConcurrentHashMap<>();
    }

    public void disconnected() {
        setStatus(ConnectionStatus.DISCONNECTED);
        reset();
    }

    public void stopped() {
        marketSubscriptionHandler = null;
        orderSubscriptionHandler = null;
        setStatus(ConnectionStatus.STOPPED);
        reset();
    }

    public MarketSubscriptionMessage getMarketResubscribeMessage() {
        if (marketSubscriptionHandler != null) {
            MarketSubscriptionMessage resub = marketSubscriptionHandler.getSubscriptionMessage();
            resub.setInitialClk(marketSubscriptionHandler.getInitialClk());
            resub.setClk(marketSubscriptionHandler.getClk());
            return resub;
        }
        return null;
    }

    public OrderSubscriptionMessage getOrderResubscribeMessage() {
        if (orderSubscriptionHandler != null) {
            OrderSubscriptionMessage resub = orderSubscriptionHandler.getSubscriptionMessage();
            resub.setInitialClk(orderSubscriptionHandler.getInitialClk());
            resub.setClk(orderSubscriptionHandler.getClk());
            return resub;
        }
        return null;
    }

    public ChangeMessageHandler getChangeHandler() {
        return changeHandler;
    }

    public void setChangeHandler(ChangeMessageHandler changeHandler) {
        if (changeHandler == null) changeHandler = new NullChangeHandler();
        this.changeHandler = changeHandler;
    }

    public SubscriptionHandler<MarketSubscriptionMessage, ChangeMessage<MarketChange>, MarketChange>
            getMarketSubscriptionHandler() {
        return marketSubscriptionHandler;
    }

    public void setMarketSubscriptionHandler(
            SubscriptionHandler<
                            MarketSubscriptionMessage, ChangeMessage<MarketChange>, MarketChange>
                    newHandler) {
        if (marketSubscriptionHandler != null) marketSubscriptionHandler.cancel();
        marketSubscriptionHandler = newHandler;
        if (marketSubscriptionHandler != null) setStatus(ConnectionStatus.SUBSCRIBED);
    }

    public SubscriptionHandler<
                    OrderSubscriptionMessage, ChangeMessage<OrderMarketChange>, OrderMarketChange>
            getOrderSubscriptionHandler() {
        return orderSubscriptionHandler;
    }

    public void setOrderSubscriptionHandler(
            SubscriptionHandler<
                            OrderSubscriptionMessage,
                            ChangeMessage<OrderMarketChange>,
                            OrderMarketChange>
                    newHandler) {
        if (orderSubscriptionHandler != null) orderSubscriptionHandler.cancel();
        orderSubscriptionHandler = newHandler;
        if (orderSubscriptionHandler != null) setStatus(ConnectionStatus.SUBSCRIBED);
    }

    public FutureResponse<ConnectionMessage> getConnectionMessage() {
        return connectionMessage;
    }

    public FutureResponse<StatusMessage> authenticate(AuthenticationMessage message)
            throws ConnectionException {
        header(message, REQUEST_AUTHENTICATION);
        return sendMessage(message, success -> setStatus(ConnectionStatus.AUTHENTICATED));
    }

    public FutureResponse<StatusMessage> heartbeat(HeartbeatMessage message)
            throws ConnectionException {
        header(message, REQUEST_HEARTBEAT);
        return sendMessage(message, null);
    }

    public FutureResponse<StatusMessage> marketSubscription(MarketSubscriptionMessage message)
            throws ConnectionException {
        header(message, REQUEST_MARKET_SUBSCRIPTION);
        SubscriptionHandler<MarketSubscriptionMessage, ChangeMessage<MarketChange>, MarketChange>
                newSub = new SubscriptionHandler<>(message, false);
        return sendMessage(message, success -> setMarketSubscriptionHandler(newSub));
    }

    public FutureResponse<StatusMessage> orderSubscription(OrderSubscriptionMessage message)
            throws ConnectionException {
        header(message, REQUEST_ORDER_SUBSCRIPTION);
        SubscriptionHandler<
                        OrderSubscriptionMessage,
                        ChangeMessage<OrderMarketChange>,
                        OrderMarketChange>
                newSub = new SubscriptionHandler<>(message, false);
        return sendMessage(message, success -> setOrderSubscriptionHandler(newSub));
    }

    /**
     * Sends a request on to the line sender
     *
     * @param message
     * @param onSuccess
     * @return
     */
    private FutureResponse<StatusMessage> sendMessage(
            RequestMessage message, Consumer<RequestResponse> onSuccess)
            throws ConnectionException {
        synchronized (sendLock) {
            int id = message.getId();
            RequestResponse requestResponse = new RequestResponse(id, message, onSuccess);

            // store a future task
            tasks.put(id, requestResponse);

            // serialize message & send
            String line = null;
            try {
                line = objectMapper.writeValueAsString(message);
            } catch (JsonProcessingException e) {
                // should never happen
                throw new ConnectionException("Failed to marshall json", e);
            }
            LOG.info("Client->ESA: " + line);

            // send line
            sendLine.sendLine(line);

            // time
            lastRequestTime = System.currentTimeMillis();

            return requestResponse.getFuture();
        }
    }

    /**
     * Sets the header on the message and assigns an id
     *
     * @param msg
     * @param op
     * @return The id
     */
    private int header(RequestMessage msg, String op) {
        int id = nextId.incrementAndGet();
        msg.setId(id);
        msg.setOp(op);
        return id;
    }

    /**
     * Processes an inbound (response) line of json
     *
     * @param line
     * @return
     */
    public ResponseMessage receiveLine(String line) throws IOException {
        // clear last response
        ResponseMessage message = objectMapper.readValue(line, ResponseMessage.class);
        lastResponseTime = System.currentTimeMillis();
        switch (message.getOp()) {
            case RESPONSE_CONNECTION -> {
                LOG.info("ESA->Client: {}", line);
                processConnectionMessage((ConnectionMessage) message);
            }
            case RESPONSE_STATUS -> {
                LOG.info("ESA->Client: {}", line);
                processStatusMessage((StatusMessage) message);
            }
            case RESPONSE_MARKET_CHANGE_MESSAGE -> {
                traceChange(line);
                processMarketChangeMessage((MarketChangeMessage) message);
            }
            case RESPONSE_ORDER_CHANGE_MESSAGE -> {
                traceChange(line);
                processOrderChangeMessage((OrderChangeMessage) message);
            }
            default -> LOG.error(
                    "ESA->Client: Unknown message type: {}, message:{}", message.getOp(), line);
        }
        return message;
    }

    private void traceChange(String line) {
        if (traceChangeTruncation != 0) {
            LOG.info(
                    "ESA->Client: {}",
                    line.substring(0, Math.min(traceChangeTruncation, line.length())));
        }
    }

    private void processOrderChangeMessage(OrderChangeMessage message) {
        ChangeMessage<OrderMarketChange> change = ChangeMessageFactory.ToChangeMessage(message);
        change = orderSubscriptionHandler.processChangeMessage(change);

        if (change != null) changeHandler.onOrderChange(change);
    }

    private void processMarketChangeMessage(MarketChangeMessage message) {
        ChangeMessage<MarketChange> change = ChangeMessageFactory.ToChangeMessage(message);
        change = marketSubscriptionHandler.processChangeMessage(change);

        if (change != null) changeHandler.onMarketChange(change);
    }

    private void processStatusMessage(StatusMessage statusMessage) {
        if (statusMessage.getId() == null) {
            // async status / status for a message that couldn't be decoded
            processUncorrelatedStatus(statusMessage);
        } else {
            RequestResponse task = tasks.get(statusMessage.getId());
            if (task == null) {
                // shouldn't happen
                processUncorrelatedStatus(statusMessage);
            } else {
                // unwind task
                task.processStatusMessage(statusMessage);
            }
        }
    }

    private void processUncorrelatedStatus(StatusMessage statusMessage) {
        LOG.error("Error Status Notification: {}", statusMessage);
        changeHandler.onErrorStatusNotification(statusMessage);
    }

    private void processConnectionMessage(ConnectionMessage message) {
        connectionMessage.setResponse(message);
        setStatus(ConnectionStatus.CONNECTED);
    }

    public static class NullChangeHandler implements ChangeMessageHandler {
        private static final Logger LOG = LoggerFactory.getLogger(NullChangeHandler.class);

        @Override
        public void onOrderChange(ChangeMessage<OrderMarketChange> change) {
            LOG.info("onOrderChange: " + change);
        }

        @Override
        public void onMarketChange(ChangeMessage<MarketChange> change) {
            LOG.info("onMarketChange: " + change);
        }

        @Override
        public void onErrorStatusNotification(StatusMessage message) {
            LOG.info("onErrorStatusNotification: " + message);
        }
    }
}
